shared api key model

a-cordier opened the pull request:
https://github.com/gravitee-io/issues/issues/6795
https://github.com/gravitee-io/issues/issues/6796
https://github.com/gravitee-io/issues/issues/6804
https://github.com/gravitee-io/issues/issues/7146
https://github.com/gravitee-io/issues/issues/7248

<!-- UI placeholder -->
🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.blob.core.windows.net/feat-shared-apikey-model/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)
<!-- UI placeholder end -->
<!-- Storybook placeholder -->
---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-ampfoeqgil.chromatic.com)
<!-- Storybook placeholder end -->

a-cordier committed:
feat: handle shared Api Key in model

- add list of subscriptions to API key model
- use new model in API key repositories
- update service and resource layer to use new model

a-cordier committed:
fix(console): sort table by name in application subscribe component

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/resources/liquibase/changelogs/v3_16_0/schema.yml:
```diff
@@ -11,3 +11,36 @@ databaseChangeLog:
                   type: nvarchar(16)
                   constraints:
                     nullable: true
+        - createTable:
```
This has to be put in the 3.17.0 schema.yml, not the 3.16.0

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/resources/liquibase/changelogs/v3_16_0/schema.yml:
```diff
@@ -11,3 +11,36 @@ databaseChangeLog:
                   constraints:
                     nullable: true
+        - createTable:
+            tableName: ${gravitee_prefix}key_subscription
```
About the table name, should it be key_subscription**s** ?

To match the other 'link' table, which are mostly at plural form.

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/mapper/KeyMapper.java:
```diff
@@ -51,6 +55,14 @@ public Key convert(ApiKeyEntity apiKeyEntity) {
+        Application application = new Application();
+        application.setId(applicationEntity.getId());
+        application.setName(applicationEntity.getName());
+        application.setApiKeyMode(Application.ApiKeyModeEnum.valueOf(applicationEntity.getApiKeyMode().name()));
```
I think the ApiKeyMode can be null in input of portal rest api.
Would it lead to a NPE here, accessing `applicationEntity.getApiKeyMode().name()` ?

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/ApiKey.java:
```diff
@@ -212,10 +214,44 @@ public void setId(String id) {
 
+    /**
+     * @deprecated
+     * Starting from 3.17 this field is kept for backward compatibility only and subscriptions should be used instead
```
Note about all deprecated fields :
I understand we have to keep those fields in database, to ensure backward compatibility with previous APIM versions.
But is there really a point to keep them in repository code ? (maybe to ensure bridge de/serialization compatibility ?)

a-cordier replied:
Yes, I kept them there for this reason but is it relevant ? 🤷 WDYT @phiz71 ?

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-test/src/test/java/io/gravitee/repository/config/mock/SubscriptionRepositoryMock.java:
```diff
@@ -123,5 +124,7 @@ void prepare(SubscriptionRepository subscriptionRepository) throws Exception {
         )
             .thenReturn(ranking);
+
+        when(subscriptionRepository.findByIdIn(argThat(ids -> ids.contains("sub1")))).thenReturn(List.of(sub1));
```
Is this mock duplicated with the one at line 86 ?

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/ApiKey.java:
```diff
@@ -44,25 +46,33 @@
+     * The subscriptions for which the Api Key is generated
      */
-    private String subscription;
+    private List<String> subscriptions;
```
During gravitee migration, what will set the list of subscriptions of the already existing api keys ?
Should we add an upgrader for that ?

a-cordier replied:
yes, the upgrader is still missing and shall be part of this PR

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/internal/key/ApiKeyMongoRepository.java:
```diff
@@ -28,12 +28,11 @@
  */
 @Repository
 public interface ApiKeyMongoRepository extends MongoRepository<ApiKeyMongo, String>, ApiKeyMongoRepositoryCustom {
+    @Query("{ 'subscriptions': ?0 }")
```
```suggestion
    @Query("{ 'subscription': ?0 }")
```

a-cordier replied:
I think this one is correct since the check should now be performed on the list of subscriptions (for a list, mongo will understand this as "where ?0 is in subscriptions")

marcambier replied:
:+1:

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/ApplicationEntity.java:
```diff
@@ -195,6 +195,10 @@ public ApiKeyMode getApiKeyMode() {
         return apiKeyMode;
     }
 
+    public boolean hasSharedMode() {
```
```suggestion
    public boolean hasSharedApiKeyMode() {
```

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-model/src/main/java/io/gravitee/rest/api/model/ApiKeyEntity.java:
```diff
@@ -29,13 +33,9 @@
-    private String subscription;
-
-    private String application;
+    private List<SubscriptionEntity> subscriptions;
```
Maybe we could initialize all those subscriptions lists to empty lists, in declaration or in builder ?
That would avoid a lot of potential NPE.

```suggestion
    private List<SubscriptionEntity> subscriptions = new ArrayList();
```

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/resources/liquibase/changelogs/v3_17_0/schema.yml:
```diff
@@ -0,0 +1,46 @@
+            baseTableName: ${gravitee_prefix}key_subscription
+            constraintName: fk_key_subscription_key
+            referencedColumnNames: id
+            referencedTableName: keys
```
The table prefix is missing, which makes liquibase migration crash if a table prefix is used.
We can detect it by running jdbc test containers; for example : `mvn test -DjdbcType=mariadb-tc -Dtest=ApiKeyRepositoryTest`

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/resources/liquibase/changelogs/v3_17_0/schema.yml:
```diff
@@ -0,0 +1,46 @@
+            baseTableName: ${gravitee_prefix}key_subscription
+            constraintName: fk_subscription_key_subscription
+            referencedColumnNames: id
+            referencedTableName: subscriptions
```
here too, missing table prefix

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-portal/gravitee-apim-rest-api-portal-rest/src/main/java/io/gravitee/rest/api/portal/rest/resource/SubscriptionKeysResource.java:
```diff
@@ -75,7 +75,7 @@ public Response revokeKeySubscription(@PathParam("subscriptionId") String subscr
         ) {
             ApiKeyEntity apiKeyEntity = apiKeyService.findByKeyAndApi(apiKey, subscriptionEntity.getApi());
-            if (apiKeyEntity.getSubscription() != null && !subscriptionId.equals(apiKeyEntity.getSubscription())) {
+            if (apiKeyEntity.getSubscriptions().stream().map(SubscriptionEntity::getId).noneMatch(subscriptionId::equals)) {
```
As SubscriptionEntity.equals() compares id ; could we simplify this expression using `apiKeyEntity.getSubscriptions().stream().noneMatch(subscriptionEntity)`

Or even, use `!apiKeyEntity.hasSubscription(subscriptionId)` ?

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/LogsServiceImpl.java:
```diff
@@ -379,7 +379,14 @@ private String getSubscription(io.gravitee.repository.log.model.ExtendedLog log)
                 ApiKeyEntity key = apiKeyService.findByKeyAndApi(log.getSecurityToken(), log.getApi());
                 if (key != null) {
-                    return key.getSubscription();
+                    // TODO
```
`TODO` comment should be removed.

Could we simplify it ? maybe by adding `Optional<SubscriptionEntity> ApiKeyEntity.getSubscriptionForApi(api)` ?

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java:
```diff
@@ -15,18 +15,23 @@
 package io.gravitee.repository.jdbc.management;
 
-import static io.gravitee.repository.jdbc.common.AbstractJdbcRepositoryConfiguration.escapeReservedWord;
+import static org.springframework.util.CollectionUtils.*;
```
Global note for all ApiKey repository modifications :
There was a lot of structural modification in ApiKeyRepository, and I think we have to add more test to ApiKeyRepositoryTest.

For example, the modification of `findById` is not tested.
We could add a test to ensure the table join is correct and subscriptions list is correctly retrieved

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java:
```diff
@@ -68,53 +90,90 @@ protected String getId(ApiKey item) {
+            }
+            return findById(apiKey.getId()).orElse(null);
+        } catch (Exception e) {
+            LOGGER.error("Failed to update api key " + apiKey.getId(), e);
```
```suggestion
            LOGGER.error("Failed to create api key " + apiKey.getId(), e);
```

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java:
```diff
@@ -68,53 +90,90 @@ protected String getId(ApiKey item) {
+            return findById(apiKey.getId()).orElse(null);
+        } catch (Exception e) {
+            LOGGER.error("Failed to update api key " + apiKey.getId(), e);
+            throw new TechnicalException("Failed to update api key " + apiKey.getId(), e);
```
```suggestion
            throw new TechnicalException("Failed to create api key " + apiKey.getId(), e);
```

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/main/java/io/gravitee/repository/jdbc/management/JdbcApiKeyRepository.java:
```diff
@@ -163,33 +237,137 @@ private boolean addClause(boolean first, StringBuilder query) {
+    private void storeSubscriptions(ApiKey key) {
+        List<String> subscriptions = key.getSubscriptions();
+
+        jdbcTemplate.update("delete from " + KEY_SUBSCRIPTION + " where key_id = ?", key.getId());
```
Not really related to this line of code, but..

We can delete subscriptions, using SubscriptionService.delete.
(Which actually not deletes Api keys).

I think that will lead to a broke FK in key_subscription ?
Maybe we have to add some key_subscription deletion during the subscription deletion process.

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -146,14 +134,40 @@ public ApiKeyEntity renew(String subscription, String customApiKey) {
+            .findFirst()
+            .map(
+                apiKey -> {
+                    apiKey.addSubscription(subscription);
```
I think we should add an audit log that indicates that a subscription has been added to a shared api key.
@mouligno WDYT ? Maybe we could plan a dedicated ticket for audit logs regarding shared api keys ?

marcambier replied:
dedicated ticket has been created

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -242,31 +241,29 @@ private void revoke(ApiKey key, boolean notify) throws TechnicalException {
+
+            // If this is not a shared API key,
+            // Get the subscription to get ending date and set key expiration date
+            if (!apiKeyEntity.getApplication().getApiKeyMode().equals(io.gravitee.rest.api.model.ApiKeyMode.SHARED)) {
```
Could we simplify this expression using `application.hasSharedApiKeyMode()` ?

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -403,36 +402,60 @@ public ApiKeyEntity updateDaysToExpirationOnLastNotification(String apiKey, Stri
+    public boolean canCreate(String apiKeyValue, SubscriptionEntity subscription) {
+        LOGGER.debug("Check if an API Key can be created for subscription {}", subscription.getId());
+
+        // TODO: make environment a parameter
```
Could we do this `TODO` if he wasn't here before ?

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -403,36 +402,60 @@ public ApiKeyEntity updateDaysToExpirationOnLastNotification(String apiKey, Stri
     @Override
-    public boolean canCreate(String apiKey, String apiId, String applicationId) {
-        LOGGER.debug("Check if an API Key can be created for api {} and application {}", apiId, applicationId);
+    public boolean canCreate(String apiKeyValue, SubscriptionEntity subscription) {
```
Maybe this one could be private now ?

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -403,36 +402,60 @@ public ApiKeyEntity updateDaysToExpirationOnLastNotification(String apiKey, Stri
+            .stream()
+            .findFirst()
+            .orElseThrow(
+                () -> new TechnicalManagementException("Unable to find subscription for API " + apiId + " and application " + applicationId)
```
Here, I think this is normal that we don't find an existing subscription.
Cause we call  `canCreate` before creating a new subscription, to ensure the custom api key is valid for it.

I think this modification causes a regression on the custom API key feature.
We can discuss about it.

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -461,21 +484,26 @@ public void delete(String apiKey) {
-                expirationDate = subscription.getEndingAt();
+            // If API key is not shared
+            // The expired date must be <= than the subscription end date
+            if (!apiKeyEntity.getApplication().getApiKeyMode().equals(io.gravitee.rest.api.model.ApiKeyMode.SHARED)) {
```
Could we simplify this expression using application.hasSharedApiKeyMode() ?

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -508,11 +536,13 @@ private static ApiKeyEntity convert(ApiKey apiKey) {
+
+        apiKeyEntity.setSubscriptions(subscriptionService.findByIdIn(apiKey.getSubscriptions()));
+
+        // TODO: make environment a parameter
```
Can we remove this `TODO` comment if it wasn't here before ?

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -527,25 +557,36 @@ private static ApiKeyEntity convert(ApiKey apiKey) {
             .expireBefore(query.getExpireBefore());
     }
 
+    // TODO: refactor to avoid querying repositories
```
Can we do this `TODO` ?

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -527,25 +557,36 @@ private static ApiKeyEntity convert(ApiKey apiKey) {
-        properties.put(APPLICATION, key.getApplication());
+        ApplicationEntity application = applicationService.findById(GraviteeContext.getCurrentEnvironment(), key.getApplication());
+
+        if (!application.getApiKeyMode().equals(io.gravitee.rest.api.model.ApiKeyMode.SHARED)) {
```
Could we simplify this expression using application.hasSharedApiKeyMode() ?

marcambier commented on gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/ApiKeyServiceImpl.java:
```diff
@@ -527,25 +557,36 @@ private static ApiKeyEntity convert(ApiKey apiKey) {
-        Map<String, Object> params = paramsBuilder.application(application).plan(plan).api(api).owner(owner).apikey(key).build();
-        notifierService.trigger(apiHook, api.getId(), params);
+
+        if (!application.getApiKeyMode().equals(ApiKeyMode.SHARED.name())) {
```
Could we simplify this expression using application.hasSharedApiKeyMode() ?

a-cordier committed:
feat: add API key subscriptions upgrader

a-cordier commented:
⚠️ The JDBC part has been tested with PostgreSQL only, due to compatibility issues between TC and Apple Silicon chips

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/internal/key/ApiKeyMongoRepositoryImpl.java:
```diff
@@ -37,38 +39,86 @@
+            apiKeyMongo.setUpdatedAt(doc.getDate("updatedAt"));
+            apiKeyMongo.setRevoked(doc.getBoolean("revoked"));
+            apiKeyMongo.setPaused(doc.getBoolean("paused"));
+            apiKeys.add(apiKeyMongo);
```
Instead of this inline mapping maybe you can use something like :
`ApiKeyMongo apiKeyMongo = mongoTemplate.getConverter().read(ApiKeyMongo.class, doc);`

marcambier commented on gravitee-apim-repository/gravitee-apim-repository-jdbc/src/test/java/io/gravitee/repository/jdbc/JdbcTestRepositoryInitializer.java:
```diff
@@ -194,7 +194,7 @@ public void tearDown() {
             (Connection con) -> {
                 for (final String table : tablesToTruncate) {
-                    jt.execute("truncate table " + escapeReservedWord(prefix + table));
+                    jt.execute("truncate table " + escapeReservedWord(prefix + table) + " CASCADE");
```
It doesn't work with all DBMS. For example, running tests with MariaDB :
```
org.springframework.jdbc.BadSqlGrammarException: StatementCallback; bad SQL grammar [truncate table `test_gio_apis` CASCADE]; nested exception is java.sql.SQLSyntaxErrorException: (conn=10) You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'CASCADE' at line 1
Caused by: java.sql.SQLSyntaxErrorException: (conn=10) You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'CASCADE' at line 1
Caused by: java.sql.SQLException: 
You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near 'CASCADE' at line 1
Query is: truncate table `test_gio_apis` CASCADE
```

a-cordier committed:
feat: handle Shared API key on gateway

Update API keys cache to use new model

a-cordier committed:
feat: add guard for API Key subscriptions in shared mode

see https://github.com/gravitee-io/issues/issues/7248

Commit:
feat: forbid usage of subscription's apiKey endpoints on shared Api Key mode

When an application uses the shared ApiKey mode,
Api key update actions (renewal, revoke, update, reactivate) have to be done using the Application's ApiKey endpoint, cause it will impact multiple subscriptions.

a-cordier committed:
fix: console error on readable api subscriptions creation

a-cordier committed:
feat: do not prompt for mode on ambiguous API key subscriptions

We cannot support multiple API key plans for the same API in shared mode,
so there is no need to prompt in that case

Commit:
feat: add Application ApiKey endpoints to handle shared api keys

- ApplicationApiKey(s)Resource handle API keys of applications using the 'shared Api key mode'.
- ApplicationSubscription(s)Resource handle API keys of applications not using the shared API key (each api key is related to a subscription)

Commit:
feat: manage shared API key in application subscriptions screen

If there are shared API keys for the application,
This add the shared API key management part at the bottom of the application subscriptions screen.

It externalize the application API key list in a dedicated component,
As it's used to manage the application API keys in :
- the application subscriptions list screen (for shared API key)
- the application subscription screen
